iree/tools/utils/image_util.c

```diff
@@ -0,0 +1,220 @@
+  IREE_TRACE_ZONE_BEGIN(z0);
+  *out_buffer_view = NULL;
+  if (element_type != IREE_HAL_ELEMENT_TYPE_SINT_8 &&
+      element_type != IREE_HAL_ELEMENT_TYPE_UINT_8) {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "element type should be i8 or u8");
+  }
+
+  iree_status_t result;
+  uint8_t* pixel_data = NULL;
+  iree_host_size_t buffer_length;
+  result = iree_tools_utils_load_pixel_data(
+      filename, shape, shape_rank, element_type, &pixel_data, &buffer_length);
+  if (!iree_status_is_ok(result)) {
+    IREE_TRACE_ZONE_END(z0);
+    return result;
+  }
+
+  iree_host_size_t element_byte = iree_hal_element_byte_count(element_type);
+  // SINT_8 and UINT_8 perform direct buffer wrap.
+  result = iree_hal_buffer_view_wrap_or_clone_heap_buffer(
+      allocator, shape, shape_rank, element_type,
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      IREE_HAL_MEMORY_ACCESS_READ, IREE_HAL_BUFFER_USAGE_ALL,
+      iree_make_byte_span((void*)pixel_data, element_byte * buffer_length),
+      iree_allocator_null(), out_buffer_view);
+
+  stbi_image_free(pixel_data);
+  IREE_TRACE_ZONE_END(z0);
+  return result;
+}
+
+iree_status_t iree_tools_utils_buffer_view_from_image_rescaled(
+    const iree_string_view_t filename, const iree_hal_dim_t* shape,
+    iree_host_size_t shape_rank, iree_hal_element_type_t element_type,
+    iree_hal_allocator_t* allocator, const float* input_range,
+    iree_host_size_t range_length, iree_hal_buffer_view_t** out_buffer_view) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  *out_buffer_view = NULL;
+  if (element_type != IREE_HAL_ELEMENT_TYPE_FLOAT_32) {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "element type should be f32");
+  }
+
+  iree_status_t result;
+  uint8_t* pixel_data = NULL;
+  iree_host_size_t buffer_length;
+  result = iree_tools_utils_load_pixel_data(
+      filename, shape, shape_rank, element_type, &pixel_data, &buffer_length);
+  if (!iree_status_is_ok(result)) {
+    IREE_TRACE_ZONE_END(z0);
+    return result;
+  }
+
+  iree_host_size_t element_byte = iree_hal_element_byte_count(element_type);
+  iree_hal_buffer_t* buffer = NULL;
+  result = iree_hal_allocator_allocate_buffer(
+      allocator,
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      IREE_HAL_BUFFER_USAGE_ALL, element_byte * buffer_length, &buffer);
+  if (!iree_status_is_ok(result)) {
+    stbi_image_free((void*)pixel_data);
+    IREE_TRACE_ZONE_END(z0);
+    return result;
+  }
+  // Need to normalize to the expected input range.
+  iree_hal_buffer_mapping_t mapped_memory;
+  result =
+      iree_hal_buffer_map_range(buffer, IREE_HAL_MEMORY_ACCESS_DISCARD_WRITE, 0,
+                                element_byte * buffer_length, &mapped_memory);
+  if (!iree_status_is_ok(result)) {
```

Comment:
in cases like these it's better to use an if(ok) pattern to avoid repeating yourself; for example:

```c
iree_status_t status = something_that_can_fail();
if (iree_status_is_ok(status)) {
  status = next_step();
}
if (iree_status_is_ok(status)) {
  status = next_step();
}
cleanup_unconditionally();
return status;
```

it's really easy for things like the releases, frees, and trace zone management to get messed up when it has to be spread throughout the function

Reply:
Done. Thanks for the suggestion.